### PR TITLE
OpenVPN Docker Image

### PR DESCRIPTION
**Pull request by goruha**
## what
* Generate SSL certs
* Store certs as k8s secrets
* Support github auth PAM
* OpenVPN

## why
* Use Github as an authorization provider for OpenVPN.
* Use OpenVPN to provide a means to access k8s wallgarden

## todo
- [ ] Documentation

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,2 +1,30 @@
 FROM kylemanna/openvpn
```
Comment (osterman) [inline Dockerfile]: Let's pin this against a specific release for stability

File: Dockerfile [hunk 2]
```diff
@@ -1,2 +1,30 @@
...
 
+RUN wget https://storage.googleapis.com/kubernetes-release/release/v1.3.6/bin/linux/amd64/kubectl
```
Comment (osterman) [inline Dockerfile]: Kubernetes version should be a parameter

File: save_secrets [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+KEY=$(cat $PKI_DIR/private/${EASYRSA_REQ_CN}.key | base64 | tr -d '\n'  )
+CA=$(cat $PKI_DIR/ca.crt | base64 | tr -d '\n'  )
+CERT=$(cat $PKI_DIR/issued/${EASYRSA_REQ_CN}.crt | base64 | tr -d '\n' )
+DH=$(cat $PKI_DIR/dh.pem | base64 | tr -d '\n' )
+TLS_AUTH=$(cat $PKI_DIR/ta.key | base64 | tr -d '\n' )
+CLIENT_KEY=$(cat $PKI_DIR/private/admin.key | base64 | tr -d '\n' )
+CLIENT=$(cat /etc/openvpn/clients/admin/admin-combined.ovpn | base64 | tr -d '\n' )
+
+
+kubectl get secrets --namespace $NAMESPACE $SECRET_NAME && ACTION=replace || ACTION=create;
```
Comment (osterman) [inline save_secrets]: This is unnecessary. Always call `kubectl apply -f -`, which will figure out if it needs to `create` or `replace`.

File: Dockerfile (hunk 1 above)
Comment (osterman) [reply Dockerfile]: Can be arg (if build args are supported there)

File: Dockerfile [hunk 4]
```diff
@@ -1,2 +1,30 @@
...
+RUN if [ ! -z $GITHUB_TOKEN ]; then \
+      set -ex \
+      && apk add --no-cache --virtual .build-deps \
+		    curl \
+		    jq \
+		  && gh-dl-release $VERSION github-pam-plugin \
+      && chmod +x github-pam-plugin \
+      && mv github-pam-plugin /bin/ \
+      && apk add ca-certificates \
+      && apk del .build-deps; \
```
Comment (osterman) [inline Dockerfile:32]:
```
else
  echo 'Cannot build without `GITHUB_TOKEN`'; \
  exit 1; \
fi
```

File: Dockerfile [hunk 5]
```diff
@@ -1,2 +1,30 @@
...
+
+ADD gh-dl-release /bin/
+RUN chmod +x /bin/gh-dl-release
+
+ARG GITHUB_TOKEN
+ARG REPO=cloudposse/github-pam
+ARG FILE=github-pam_linux_386
+ARG VERSION=0.4
+
+RUN if [ ! -z $GITHUB_TOKEN ]; then \
```
Comment (osterman) [inline Dockerfile:19]: `! -z` can simply be written as `-n`

Comment (goruha) [reply Dockerfile:19]: Done

File: Dockerfile (hunk 4 above)
Comment (goruha) [reply Dockerfile:32]: Done

File: Dockerfile (hunk 5 above)
Comment (goruha) [reply Dockerfile:19]: -n works different then ! -z
! -z return true only of env var exists and not empty

File: save_secrets (hunk 3 above)
Comment (goruha) [reply save_secrets]: Done

File: Dockerfile (hunk 2 above)
Comment (goruha) [reply Dockerfile]: Done

File: Dockerfile (hunk 1 above)
Comment (goruha) [reply Dockerfile]: it does not work this way

````
goruha@goruha-laptop-hp ~/projects/cloudposse/docker-openvpn (initial●)$ docker build ./                                                                                            [system]
Sending build context to Docker daemon 229.9 kB
Step 1 : ARG K8S_VERSION=v1.3.6
Please provide a source image with `from` prior to commit
goruha@goruha-laptop-hp ~/projects/cloudposse/docker-openvpn (initial●)$ docker build ./                                                                                            [system]
Sending build context to Docker daemon 229.9 kB
Step 1 : FROM kylemanna/openvpn:$BASE_IMAGE_TAG
Error parsing reference: "kylemanna/openvpn:$BASE_IMAGE_TAG" is not a valid repository/tag
````

Comment (osterman) [reply Dockerfile]: ok - thanks for verifying. let's then pin it to a tag'd release.

File: Dockerfile [hunk 6]
```diff
@@ -1,2 +1,33 @@
...
+
+RUN wget https://storage.googleapis.com/kubernetes-release/release/$K8S_VERSION/bin/linux/amd64/kubectl
+RUN chmod +x kubectl
+RUN mv kubectl /usr/local/bin/
+
+ADD gh-dl-release /bin/
+RUN chmod +x /bin/gh-dl-release
+
+ARG GITHUB_TOKEN
+ARG REPO=cloudposse/github-pam
```
Comment (osterman) [inline Dockerfile:9]: Think these should be prefixed with `GITHUB_PAM_`

File: Dockerfile [hunk 7]
```diff
@@ -1,2 +1,63 @@
-FROM kylemanna/openvpn
+FROM kylemanna/openvpn:latest
 
+ARG K8S_VERSION=v1.3.6
```
Comment (osterman) [inline Dockerfile]: I think we should bump this to the current release.

File: rootfs/lib/security/github/pam_script_auth [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+github-pam-plugin $1
```
Comment (osterman) [inline rootfs/lib/security/github/pam_script_auth]: this worries me. the `$1` unescaped.

File: rootfs/lib/security/github/pam_script_auth [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+PAM_SCRIPT_STATUS=$?
+export PAM_SCRIPT_STATUS
```
Comment (osterman) [inline rootfs/lib/security/github/pam_script_auth]: no need to export... its not used by any subshells.

File: rootfs/init.sh [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+sed -i "s|{{GITHUB_PAM_TEAM}}|$GITHUB_PAM_TEAM|g" /etc/pam.d/openvpn
```
Comment (osterman) [inline rootfs/init.sh]: prefer to use `envsubst` command

File: Dockerfile [hunk 11]
```diff
@@ -1,2 +1,63 @@
...
+      && apk del .build-deps; \
+    else \
+      echo '`GITHUB_TOKEN` required for fetching github-pam-plugin'; \
+      exit 1; \
+    fi
+
+RUN set -ex \
+      && apk update \
+      && apk add linux-pam \
+      && apk add --no-cache --virtual .build-deps \
```
Comment (osterman) [inline Dockerfile:41]: ...nice trick with the `--virtual .build-deps`

File: rootfs/bin/save_secrets [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+TYPE=${TYPE:-Opaque}
+
+if [ -z "$SECRET_NAME" ]; then
+    echo "ERROR: Secret name is empty or unset"
+    exit 1
+fi
+
+PKI_DIR=/etc/openvpn/pki
+
+KEY=$(cat $PKI_DIR/private/${EASYRSA_REQ_CN}.key | base64 | tr -d '\n'  )
```
Comment (osterman) [inline rootfs/bin/save_secrets]: can write simply 
```
KEY=$(base64 $PKI_DIR/private/${EASYRSA_REQ_CN}.key | tr -d '\n')
```

File: rootfs/bin/gh-dl-release [hunk 13]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
```
Comment (osterman) [inline rootfs/bin/gh-dl-release]: It pains me we keep copying this into repos. Can't we just create our own `gh-dl-release` repo, and `ADD` the GitHub raw URL?

File: Dockerfile [hunk 14]
```diff
@@ -1,2 +1,63 @@
...
+RUN if [ ! -z $GITHUB_TOKEN ]; then \
+      set -ex \
+      && echo "http://dl-cdn.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories \
+      && apk update \
+      && apk add --no-cache --virtual .build-deps \
+		    curl \
+		    jq \
+		    linux-pam \
+		    pamtester \
+		  && gh-dl-release $VERSION github-pam-plugin \
```
Comment (osterman) [inline Dockerfile]: whitespace for line #26 should be same for #27?

File: Dockerfile (hunk 7 above)
Comment (goruha) [reply Dockerfile]: Done

File: Dockerfile (hunk 11 above)
Comment (goruha) [reply Dockerfile:41]: yea... nice to clear dev packages in alpine.
no sure this is possible in other package managers like apt

File: rootfs/lib/security/github/pam_script_auth (hunk 8 above)
Comment (goruha) [reply rootfs/lib/security/github/pam_script_auth]: Done

File: rootfs/lib/security/github/pam_script_auth (hunk 9 above)
Comment (goruha) [reply rootfs/lib/security/github/pam_script_auth]: Done

File: rootfs/init.sh (hunk 10 above)
Comment (goruha) [reply rootfs/init.sh]: Done

File: rootfs/bin/save_secrets (hunk 12 above)
Comment (goruha) [reply rootfs/bin/save_secrets]: Done

File: rootfs/bin/gh-dl-release (hunk 13 above)
Comment (goruha) [reply rootfs/bin/gh-dl-release]: Done

File: Dockerfile (hunk 11 above)
Comment (osterman) [reply Dockerfile:41]: Yea, not seen that before. Alpine is definitely catering to the docker generation.

File: Dockerfile (hunk 14 above)
Comment (goruha) [reply Dockerfile]: Done

**Review by osterman — APPROVED** [review]

**Pull request merged**